### PR TITLE
Criu

### PR DESCRIPTION
**Pull request by sebratton**
Add logging and changes to server launchl script

### CODE REVIEW

File: dev/com.ibm.ws.kernel.boot.ws-server/publish/bin/server [hunk 1]
```diff
@@ -69,7 +69,8 @@
 #              a z/OS environment.
 #
 ###############################################################################
-
+export LD_LIBRARY_PATH=/usr/lib/x86_64-linux-gnu
+export LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libcriu.so
```
Comment (tjwatson) [inline dev/com.ibm.ws.kernel.boot.ws-server/publish/bin/server:73]: Will need to remember to remove this once we have the JVM functionality to do checkpoint.

File: dev/com.ibm.ws.kernel.boot.ws-server/publish/bin/server [hunk 2]
```diff
@@ -744,12 +745,12 @@ readServerEnv()
         # Only accept alphanumeric variable names to avoid eval complexities.
         if var=`safeEcho "$line" | sed -e 's/^\([a-zA-Z0-9_][a-zA-Z0-9_]*\)=.*/\1/'`; then
            case $var in
-	   *=*)
+	         *=*)
```
Comment (tjwatson) [inline dev/com.ibm.ws.kernel.boot.ws-server/publish/bin/server]: The format here looks off

File: dev/com.ibm.ws.kernel.boot.ws-server/publish/bin/server [hunk 3]
```diff
@@ -744,12 +745,12 @@ readServerEnv()
...
                SERVER_ENV_SETUP_FAILURE="${var} ${SERVER_ENV_SETUP_FAILURE}" ;;
            *)
               extractValueAndEscapeForEval "${line}"
               eval "${var}=${escapeForEvalResult}; export ${var}" ;;
-	   esac
+	         esac
```
Comment (tjwatson) [inline dev/com.ibm.ws.kernel.boot.ws-server/publish/bin/server]: Format looks off

File: dev/com.ibm.ws.kernel.boot.ws-server/publish/bin/server [hunk 4]
```diff
@@ -826,10 +827,18 @@ javaCmd()
     if [ -n "${option}" ]; then
       if [ -z "${option##-D*}" ] || [ -z "${option##-X*}" ]; then
         JVM_OPTIONS_QUOTED="$JVM_OPTIONS_QUOTED '$option'"
-      elif [ -z "$REMAINING_ARGS" ] ; then
-        REMAINING_ARGS="$option"
       else
-        REMAINING_ARGS="$REMAINING_ARGS $option"
+        echo OPTION : ${option}
```
Comment (tjwatson) [inline dev/com.ibm.ws.kernel.boot.ws-server/publish/bin/server]: Is this just for debugging?

File: dev/com.ibm.ws.kernel.boot.ws-server/publish/bin/server [hunk 5]
```diff
@@ -826,10 +827,18 @@ javaCmd()
...
+        case ${option} in 
+        --checkpoint=*)
+           CHECKPOINT=${option##--checkpoint=}; 
+           echo CHECKPOINT : ${CHECKPOINT} ;;
```
Comment (tjwatson) [inline dev/com.ibm.ws.kernel.boot.ws-server/publish/bin/server]: dubugging?

File: dev/com.ibm.ws.kernel.boot.ws-server/publish/bin/server [hunk 6]
```diff
@@ -826,10 +827,18 @@ javaCmd()
     if [ -n "${option}" ]; then
       if [ -z "${option##-D*}" ] || [ -z "${option##-X*}" ]; then
         JVM_OPTIONS_QUOTED="$JVM_OPTIONS_QUOTED '$option'"
-      elif [ -z "$REMAINING_ARGS" ] ; then
-        REMAINING_ARGS="$option"
       else
-        REMAINING_ARGS="$REMAINING_ARGS $option"
+        echo OPTION : ${option}
+        case ${option} in
```
Comment (tjwatson) [inline dev/com.ibm.ws.kernel.boot.ws-server/publish/bin/server]: Why have a case for a single option value here?

File: dev/com.ibm.ws.kernel.boot.ws-server/publish/bin/server [hunk 7]
```diff
@@ -841,6 +850,20 @@ javaCmd()
...
+         echo remove existing snapshot
+         rm -r ${SERVER_OUTPUT_DIR}/workarea/snapshot
+         #TODO check for error removing existing snapshot
+       fi
+       #TODO remove setting of LD_PRELOAD when running dump as root no longer required
+       LD_LIBRARY_PATH=/usr/lib/x86_64-linux-gnu${LD_LIBRARY_PATH:+:$LD_LIBRARY_PATH}
+       LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libcriu.so
+     fi
+  fi
+  echo s0
```
Comment (tjwatson) [inline dev/com.ibm.ws.kernel.boot.ws-server/publish/bin/server]: are the echo calls above meant for debug or are they meant to stay for production?

File: dev/com.ibm.ws.kernel.boot.ws-server/publish/bin/server [hunk 8]
```diff
@@ -841,6 +850,20 @@ javaCmd()
...
+     if [ -n "${CHECKPOINT}" ]; then                                               
+       echo looking for file: ${SERVER_OUTPUT_DIR}/workarea/snapshot
+       if [ -d "${SERVER_OUTPUT_DIR}/workarea/snapshot" ]; then 
+         echo remove existing snapshot
+         rm -r ${SERVER_OUTPUT_DIR}/workarea/snapshot
+         #TODO check for error removing existing snapshot
+       fi
+       #TODO remove setting of LD_PRELOAD when running dump as root no longer required
+       LD_LIBRARY_PATH=/usr/lib/x86_64-linux-gnu${LD_LIBRARY_PATH:+:$LD_LIBRARY_PATH}
+       LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libcriu.so
```
Comment (tjwatson) [inline dev/com.ibm.ws.kernel.boot.ws-server/publish/bin/server]: Why is this set here and in some export calls at the beginning of the script?

File: dev/com.ibm.ws.kernel.boot.ws-server/publish/bin/server [hunk 9]
```diff
@@ -1054,21 +1077,37 @@ serverCmd()
...
+        echo Verify/wait for the process to start 
+        echo javaCmd '' "${SERVER_NAME}" --pid="${PID}" --status:start "$@"
+        javaCmd '' "${SERVER_NAME}" --pid="${PID}" --status:start "$@"
+        rc=$?
+        # write the pid file if return is OK or UNKNOWN
+        if [ $rc = 0 -o $rc = $SERVER_UNKNOWN_STATUS ]; then
+          safeEcho "${PID}" > "${X_PID_FILE}"
+        fi        
+      fi    
+    fi
```
Comment (tjwatson) [inline dev/com.ibm.ws.kernel.boot.ws-server/publish/bin/server:1088]: I assume the echo calls above are just for debugging.

File: dev/com.ibm.ws.kernel.boot.ws-server/publish/bin/server [hunk 10]
```diff
@@ -1327,6 +1366,33 @@ case "$ACTION" in
...
+    
+    # no good way to check for criu support since fats startu is imperitive
+    #   just let criu invocation fail with a hopefully informative error message.
+    if [ ! -d "${SERVER_OUTPUT_DIR}/workarea/snapshot" ]; then
+      echo Snapshot image not found at ${SERVER_OUTPUT_DIR}/workarea/snapshot.
+      exit 2
+    fi  
+    # set LD_LIBRARY path
+    LD_LIBRARY_PATH=/usr/lib/x86_64-linux-gnu${LD_LIBRARY_PATH:+:$LD_LIBRARY_PATH}
+    LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libcriu.so
```
Comment (tjwatson) [inline dev/com.ibm.ws.kernel.boot.ws-server/publish/bin/server]: Do you need to set these LD vars for the restore side?

File: dev/com.ibm.ws.kernel.boot.ws-server/publish/bin/server (hunk 1 above)
Comment (sebratton) [reply dev/com.ibm.ws.kernel.boot.ws-server/publish/bin/server:73]: Agreed

**Pull request merged**